### PR TITLE
generate asymmetric keys step

### PR DESCRIPTION
**Pull request by stevenwarejones**
This adds in generate serialized rlwe keys step. It also now only has the daemon instantiate a cryptor class if its needed for that particular step.

---
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/149)

### CODE REVIEW

**Review by duliomatos1 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/149)* status: 0 of 21 files reviewed, 1 unresolved discussion (waiting on @efoxepstein and @stevenwarejones)

___
*[src/main/kotlin/org/wfanet/panelmatch/client/exchangetasks/GenerateAsymmetricKeysTask.kt, line 32 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/149#-MlLXap0Ezatq5iuH6oy:-MlLXap0Ezatq5iuH6oz:b97mt76) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/b37388f5d5df7470a04b8df73424ad6e545e0a49/src/main/kotlin/org/wfanet/panelmatch/client/exchangetasks/GenerateAsymmetricKeysTask.kt#L32)):*
> ```Kotlin
> 
>   override suspend fun execute(input: Map<String, VerifiedBlob>): Map<String, Flow<ByteString>> {
>     logger.addToTaskLog("Executing generate asymmetric keys: $generateKeys")
> ```

How will `$generateKeys` be printed here, being a function?

**Review by efoxepstein — CHANGES_REQUESTED** [review]
> Reviewed 21 of 21 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/149)* status: all files reviewed, 7 unresolved discussions (waiting on @duliomatos1 and @stevenwarejones)

___
*[src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt, line 87 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/149#-MlLVi8mE7mm1rs2gRW4:-MlLVi8mE7mm1rs2gRW5:bsvrewo) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/b37388f5d5df7470a04b8df73424ad6e545e0a49/src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt#L87)):*
> ```Kotlin
>     val exchangeTaskMapper =
>       ExchangeTaskMapperForJoinKeyExchange(
>         getDeterministicCommutativeCryptor = fun() = JniDeterministicCommutativeCipher(),
> ```

`... = ::JniDeterministicCommutativeCipher`
___
*[src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt, line 89 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/149#-MlLVmQ754tvwhgJmfIj:-MlLVmQ754tvwhgJmfIk:b-b5iih9) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/b37388f5d5df7470a04b8df73424ad6e545e0a49/src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt#L89)):*
> ```Kotlin
>         getDeterministicCommutativeCryptor = fun() = JniDeterministicCommutativeCipher(),
>         getPrivateMembershipCryptor =
>           fun(serializedParameters: ByteString) = JniPrivateMembershipCryptor(serializedParameters),
> ```

`... = ::JniPrivateMembershipCryptor`
___
*[src/main/kotlin/org/wfanet/panelmatch/client/exchangetasks/GenerateAsymmetricKeysTask.kt, line 27 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/149#-MlLW2vQDWGIxdUNqgYv:-MlLW2vQDWGIxdUNqgYw:b-eqjj0p) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/b37388f5d5df7470a04b8df73424ad6e545e0a49/src/main/kotlin/org/wfanet/panelmatch/client/exchangetasks/GenerateAsymmetricKeysTask.kt#L27)):*
> ```Kotlin
> class GenerateAsymmetricKeysTask(
>   private val generateKeys: () -> AsymmetricKeys,
>   private val privateKeyLabel: String = "private-key",
> ```

Nit: I would say either remove the parameters totally (and move the labels to constants) or remove the defaults and force the caller to specify them. I feel like as is, it feels too flexible/unpredictable.

(And in general I think all of us have probably overused default parameter values)
___
*[src/main/kotlin/org/wfanet/panelmatch/client/exchangetasks/GenerateSymmetricKeyTask.kt, line 30 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/149#-MlLY7_q0g7dznz4eGJI:-MlLY7_r6SphHxX1ucM1:b-9rtnaq) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/b37388f5d5df7470a04b8df73424ad6e545e0a49/src/main/kotlin/org/wfanet/panelmatch/client/exchangetasks/GenerateSymmetricKeyTask.kt#L30)):*
> ```Kotlin
> 
>   override suspend fun execute(input: Map<String, VerifiedBlob>): Map<String, Flow<ByteString>> {
>     logger.addToTaskLog("Executing generate symmetric key: $generateKey")
> ```

Do you have any worries that `generateKey.toString()` could accidentally leak a salt or something?
___
*[src/test/kotlin/org/wfanet/panelmatch/client/exchangetasks/ExchangeTaskMapperForJoinKeyExchangeTest.kt, line 37 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/149#-MlLYXpV8SKa-1bz_uNL:-MlLYXpV8SKa-1bz_uNM:b-7ropln) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/b37388f5d5df7470a04b8df73424ad6e545e0a49/src/test/kotlin/org/wfanet/panelmatch/client/exchangetasks/ExchangeTaskMapperForJoinKeyExchangeTest.kt#L37)):*
> ```Kotlin
>   private val exchangeTaskMapper =
>     ExchangeTaskMapperForJoinKeyExchange(
>       getDeterministicCommutativeCryptor = fun() = FakeDeterministicCommutativeCipher(),
> ```

ditto
___
*[src/test/kotlin/org/wfanet/panelmatch/client/exchangetasks/ExchangeTaskMapperForJoinKeyExchangeTest.kt, line 38 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/149#-MlLYYdp8ACIfF5z73NQ:-MlLYYdp8ACIfF5z73NR:b-89a4l7) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/b37388f5d5df7470a04b8df73424ad6e545e0a49/src/test/kotlin/org/wfanet/panelmatch/client/exchangetasks/ExchangeTaskMapperForJoinKeyExchangeTest.kt#L38)):*
> ```Kotlin
>     ExchangeTaskMapperForJoinKeyExchange(
>       getDeterministicCommutativeCryptor = fun() = FakeDeterministicCommutativeCipher(),
>       getPrivateMembershipCryptor =
> ```

Ditto

**Review by efoxepstein — APPROVED** [review]
> Reviewed 7 of 7 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/149)* status: all files reviewed, 2 unresolved discussions (waiting on @duliomatos1 and @stevenwarejones)

___
*[src/main/kotlin/org/wfanet/panelmatch/client/exchangetasks/GenerateSymmetricKeyTask.kt, line 30 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/149#-MlLY7_q0g7dznz4eGJI:-MlMzmhQ7r7TnLFUVTw1:b-f8pjv9) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/b37388f5d5df7470a04b8df73424ad6e545e0a49/src/main/kotlin/org/wfanet/panelmatch/client/exchangetasks/GenerateSymmetricKeyTask.kt#L30)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

it doesn't print the passed variable. How would it leak it in your mind?
</blockquote></details>

Is there a guarantee in the Kotlin spec that `KFunction1` or whatever the class of `generateKey` is won't have a `toString` that prints out closed-over variables?

Offhand, do you know what the result of `Foo("some-secret")::foo.toString()` is given `data class Foo(val secret: String) { fun foo(): ByteString {...} }`?

Basically, it just seems like a potential way to leak a secret.

**Review by duliomatos1 — APPROVED** [review]
> Reviewed 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/149)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @stevenwarejones)
